lib/unifex/native_code_generator.ex

```diff
@@ -340,13 +357,46 @@ defmodule Unifex.NativeCodeGenerator do
       enif_release_resource(state);
     }
 
+    void unifex_keep_state(UnifexEnv * env, #{state_type} state) {
+      UNIFEX_UNUSED(env);
+      enif_keep_resource(state);
+    }
+
     static void destroy_state(ErlNifEnv* env, void* value) {
       #{state_type} state = (#{state_type}) value;
       #{generate_unifex_env()}
       handle_destroy_state(unifex_env, state);
     }
+    """
+  end
 
-    static int load(ErlNifEnv* env, void** priv_data, ERL_NIF_TERM load_info) {
+  defp generate_nif_lifecycle_callbacks_declarations(callbacks) do
+    callbacks
+    |> Enum.map_join("\n", fn
+      {:load, fun_name} ->
+        ~g"int #{fun_name}(UnifexEnv * env, void ** priv_data);"
+
+      {:upgrade, fun_name} ->
+        ~g"int #{fun_name}(UnifexEnv * env, void ** priv_data, void **old_priv_data);"
+
+      {:unload, fun_name} ->
+        ~g"void #{fun_name}(UnifexEnv * env, void * priv_data);"
+    end)
+  end
+
+  defp generate_nif_lifecycle_callbacks(nil, _callbacks) do
+    ~g""
+  end
+
+  defp generate_nif_lifecycle_callbacks(_module, callbacks) do
+    load_result =
+      case callbacks[:load] do
```

Comment:
all these cases look too verbose and violate DRY

Reply:
I disagree. Being verbose is not a flaw (and it could have been less verbose if not for a formatter) and the only common thing for the cases is changing `nil` to `""` and passing a name to a code template in 2 out 3 cases. Not worth the effort.